rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,258 @@
+---
+authors: Tim Ross (tim.ross@goteleport.com)
+state: draft
+---
+
+# RFD 126 - Backend Migrations
+
+## Required Approvers
+
+- Engineering @zmb3 && (@fspmarshall || @espadollini )
+- Product: ( @klizhentas || @russjones )
+
+## What
+
+A more reliable and robust mechanism to perform backend migrations.
+
+## Why
+
+The upgrade process requires scaling Auth down to a single instance to ensure that migrations are only performed once as well as preventing a new version and an old version from operating on the same keys with different schema. This is cumbersome and makes the upgrade process a pain point for cluster admins. Scaling Auth down and back up also results in an [uneven load](https://github.com/gravitational/teleport/issues/7029) which can cause connectivity issues and backend latency that can result in a thundering herd.
+
+Migrations to date have existed as a function that is called during Auth initialization. The migration function is left as is for a release or two and is then deleted. This makes tracking when and which migrations have been applied difficult and prevents skipping any major versions when upgrading since there are no guarantees that the latest version will include all the migrations between versions.
```

Comment:
You could also link to https://github.com/gravitational/teleport/issues/12365 which highlights the difficulty of tracking migrations.